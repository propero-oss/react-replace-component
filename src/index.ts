import { createStore, ReadOnlyStore } from "@propero/easy-store";
import { useEffect, useState } from "react";

type Replacement<T = unknown> = [T, T];

interface ComponentData {
  namespace: string;
  components: Record<string, Replacement[]>;
}

export type MappedComponents<T> = {
  [Key in keyof T as `${Key & string}Ex`]: T[Key];
};

export interface ComponentStore extends ReadOnlyStore<ComponentData> {
  registerReplacement<T>(component: T, replacement: T, ns?: string): void;
  getComponent<T>(component: T): T;
  useComponent<T>(component: T): T;
  useComponents<T>(components: T): MappedComponents<T>;
  setNamespace(namespace: string): void;
  getNamespace(): string;
}

export function createComponentStore(): ComponentStore {
  const { sub, unsub, update, getValue } = createStore<ComponentData>({
    namespace: "default",
    components: {},
  });

  function registerReplacement<T>(component: T, replacement: T, ns: string = "default"): void {
    update(({ components, namespace }) => {
      let replacements = (components[ns] ??= []);
      if (replacements.find(([original]) => original === component))
        throw new Error(`Cannot register one component multiple times in one namespace.`);
      replacements = [...replacements, [component, replacement]];
      components = { ...components, [ns]: replacements };
      return { components, namespace };
    });
  }

  function getComponent<T>(component: T, { components, namespace }: ComponentData = getValue()): T {
    return (
      (components[namespace]?.find(([original]) => original === component)?.[1] as T) ?? component
    );
  }

  function useComponent<T>(component: T): T {
    const [comp, setComp] = useState<T>(() => getComponent(component));

    useEffect(() => {
      function handler(newVal: ComponentData, oldVal: ComponentData): void {
        const newComp = getComponent(component, newVal);
        const oldComp = getComponent(component, oldVal);
        if (newComp !== oldComp) setComp(() => newComp);
      }
      sub(handler);
      return () => unsub(handler);
    }, [component]);

    return comp;
  }

  function useComponents<T>(components: T): MappedComponents<T> {
    return Object.fromEntries(
      Object.entries(components).map(([name, component]) => [`${name}Ex`, useComponent(component)])
    ) as any;
  }

  function setNamespace(namespace: string): void {
    update(({ components }) => ({ namespace, components }));
  }

  function getNamespace(): string {
    return getValue().namespace;
  }

  return {
    sub,
    unsub,
    getValue,
    registerReplacement,
    getComponent,
    useComponent,
    useComponents,
    setNamespace,
    getNamespace,
  };
}
