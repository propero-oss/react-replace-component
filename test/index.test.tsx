import React, { ReactElement } from "react";
import { ComponentStore, createComponentStore } from "src/index";
import renderer, { act, ReactTestRenderer, TestRendererOptions } from "react-test-renderer";

function A(): JSX.Element {
  return <div className="a" />;
}

function B(): JSX.Element {
  return <div className="b" />;
}

function C(): JSX.Element {
  return <div className="c" />;
}

function Playground({ store }: { store: ComponentStore }): JSX.Element {
  const { useComponent, useComponents } = store;
  const AEx = useComponent(A);
  const { BEx, CEx } = useComponents({ B, C });
  return (
    <>
      <AEx />
      <BEx />
      <CEx />
    </>
  );
}

function create(nextElement: ReactElement, options?: TestRendererOptions): ReactTestRenderer {
  let component: ReactTestRenderer;
  act(() => {
    component = renderer.create(nextElement, options);
  });
  return component!;
}

describe("index.ts", () => {
  let store: ComponentStore;
  beforeEach(() => (store = createComponentStore()));

  it("should pass through components that are not replaced", () => {
    const rendered = create(<Playground store={store} />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it("should pass through components not registered in current namespace", () => {
    const rendered = create(<Playground store={store} />);
    store.registerReplacement(A, B, "foo");
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it("should replace components replaced in registered namespace", () => {
    const rendered = create(<Playground store={store} />);
    expect(rendered.toJSON()).toMatchSnapshot();
    act(() => store.registerReplacement(A, B));
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it("should restore components no longer replaced on namespace change", () => {
    const rendered = create(<Playground store={store} />);
    expect(rendered.toJSON()).toMatchSnapshot();
    act(() => store.registerReplacement(A, B));
    expect(rendered.toJSON()).toMatchSnapshot();
    act(() => store.setNamespace("bar"));
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it("should return the current namespace", () => {
    expect(store.getNamespace()).toEqual("default");
    store.setNamespace("foo");
    expect(store.getNamespace()).toEqual("foo");
  });

  it("should throw for multiple registrations of the same component in the same namespace", () => {
    store.registerReplacement(A, B);
    expect(() => store.registerReplacement(A, C)).toThrow();
  });
});
