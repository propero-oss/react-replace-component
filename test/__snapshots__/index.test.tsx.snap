// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.ts should pass through components not registered in current namespace 1`] = `
Array [
  <div
    className="a"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;

exports[`index.ts should pass through components that are not replaced 1`] = `
Array [
  <div
    className="a"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;

exports[`index.ts should replace components replaced in registered namespace 1`] = `
Array [
  <div
    className="a"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;

exports[`index.ts should replace components replaced in registered namespace 2`] = `
Array [
  <div
    className="b"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;

exports[`index.ts should restore components no longer replaced on namespace change 1`] = `
Array [
  <div
    className="a"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;

exports[`index.ts should restore components no longer replaced on namespace change 2`] = `
Array [
  <div
    className="b"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;

exports[`index.ts should restore components no longer replaced on namespace change 3`] = `
Array [
  <div
    className="a"
  />,
  <div
    className="b"
  />,
  <div
    className="c"
  />,
]
`;
